### PR TITLE
Move WP_REST_Block_Navigation_Areas_Controller into Core

### PR DESCRIPTION
**Pull request by anton-vlasenko**
This PRs moves the WP_REST_Block_Navigation_Areas_Controller controller into Core.
Also, it adds a functional test for it.

Trac ticket: https://core.trac.wordpress.org/ticket/54393

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [inline src/wp-includes/block-template-areas.php:19]: Should we add a `wp_` prefix to these functions? (I am never sure.)

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [inline src/wp-includes/block-template-areas.php:16]:
```suggestion
 *
 * @since 5.9.0
 *
```

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [inline src/wp-includes/block-template-areas.php:1]: Not sure about the naming of this file. Should it be `wp-includes/navigation-areas.php`? Or, alternatively, should this functionality go into the existing `wp-includes/block-template.php?`

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [inline src/wp-includes/block-template-areas.php:35]:
```suggestion
 *
 * @since 5.9
 *
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (noisysocks) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:11]: I won't spam but all these doc comments need a `@since`.

File: src/wp-includes/block-template-areas.php
Comment (TimothyBJacobs) [reply src/wp-includes/block-template-areas.php:1]: Yeah this is small enough I'd just put it in `block-template.php`.

File: src/wp-includes/block-template-areas.php
Comment (TimothyBJacobs) [reply src/wp-includes/block-template-areas.php:19]: I think it's probably fine without, but a better person to ask might be @peterwilsoncc.

File: src/wp-includes/block-template-areas.php
Comment (TimothyBJacobs) [inline src/wp-includes/block-template-areas.php:25]: This should be attached to some hook like `setup_theme` that is hooked in in `default-filters.php`.

File: src/wp-includes/block-template-areas.php
Comment (TimothyBJacobs) [inline src/wp-includes/block-template-areas.php:27]: I think these should probably be localized.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:6]:
```suggestion
 * @package WordPress
```

These tags don't get aligned, https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#6-file-headers.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:41]: `allow_batch` doesn't make sense here since this is a `GET` endpoint which we don't support batching for. Instead this should be in the route registration for the second set of routes.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:81]:
```suggestion
	public function get_items_permissions_check( $request ) {
```

Don't need these in Core. We should probably this lint in GB tbh.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:85]:
```suggestion
				__( 'Sorry, you are not allowed to view navigation areas.' ),
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:105]: If we can make this just a list, ie a numeric instead of associative array that'd be preferable. The REST API assumes that a collection response is a list, we break this in a couple of places, but we should try to avoid introducing new breaks.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:85]: This applies to multiple lines.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:98]:
```suggestion
	 * @param WP_REST_Request $request Full details about the request.
	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
```

No blank line between `@param` and `@return`. https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:183]:
```suggestion
	protected function get_navigation_area_object( $name ) {
```

We should give extending classes the flexibility here I think.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:234]: These need to be aligned, https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:245]: Let's store this in `$this→schema` and return from there if it exists. This saves us running string localization multiple times which can stack up.

File: tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php:9]: Naming for this should follow https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes, see the `Tests_REST_WpRestUrlDetailsController` class for an example. (I know other classes don't follow this pattern, a WIP!)

File: tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php:90]:
```suggestion
		$this->markTestAsSkipped();
```

File: tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php
Comment (TimothyBJacobs) [inline tests/phpunit/tests/rest-api/rest-block-navigation-areas-controller.php:40]:
```suggestion
		$request = new WP_REST_Request( 'GET', '/wp/v2/block-navigation-areas' );
```

I'd just type out the request methods here.

**Review by TimothyBJacobs — CHANGES_REQUESTED** [review]
Left some feedback, mostly minor. Most major thing is changing the return type to be strictly a list instead of an object. If that requires JS changes, we can wait until after merge to address it. But we should make sure that gets cleaned up before b1.

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [reply src/wp-includes/block-template-areas.php:1]: On second thought I think let's rename to `navigation-areas.php`. I looked at `block-template-utils.php` and `block-template.php` but neither really feel right to me. Looking at `lib/navigation.php` in Gutenberg, there's some more things we probably will want to move over into this file.

File: src/wp-includes/block-template-areas.php
Comment (noisysocks) [reply src/wp-includes/block-template-areas.php:19]: I am thinking without the prefix as that matches `register_sidebar`.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php
Comment (noisysocks) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-block-navigation-areas-controller.php:105]: I think it's safe to change because the frontend is calling this endpoint via `core-data` which should handle arrays just fine.

**Comment by noisysocks** [comment]
Thanks @anton-vlasenko and @TimothyBJacobs!

**Pull request closed without merging**
